Add ability to add to/customize context menu

RobAley opened the pull request:
I see that the ability to add custom items to the context menu is in progress for version 6, however I'm guessing v6 is still some way away. 

This pull request introduces a simple way to allow customization of the context menu in the current version. It simply lets the user register an "onCreateMenu" function in options to which the contents of the menu "items" variable (plus reference to the current node) are passed before the menu is created from them.

I currently use it internally to add a custom menu item to display the current path as a JSON Pointer path (by pushing a new item onto items), and to alter the styling & text of some of the standard menu items depending on the current path (by altering existing item in items).

If this is something you want to incorporate, I can work up an example for the examples folder and additions to the api docs.

josdejong commented:
Thanks @RobAley! Yes makes sense not to await v6.

Some feedback:

1. Can you describe the new feature in the docs? Also, an example demonstrating how to use it would be great.
2. Right now you pass `node` as second argument, but so far the nodes are internal and undocumented API. Maybe we can instead pass an object containing the path of the current node, similar to `onEditable` and `onEvent`  and some other callbacks work. What do you think?

RobAley commented:
I agree with returning path rather than node, I can always find the node from the path if needed.

I've changed it to return the path rather than node. I've also added it to appendNodeFactory.js and treemode.js to capture (i think) everywhere a tree's context menu is created.

I've also added an example and put it in the docs.

Let me know what you think. Happy to change anything (or for you to do so).

josdejong commented:
wow, that's really nice @RobAley !! I love the example, it clearly demonstrates the power of customizing the menu, thanks.

Thanks for adjusting the API. I would love to see one more change there: change the second argument of `onCreateMenu` from being the path to being an object containing the path, like `onCreateMenu(items, node)` where node is an object `{ path: [...] }`. This will make it consistent with other functions like `onEditable`, and it makes it easy to expand the API in the future if needed by adding more properties to the `node` object.

RobAley commented:
Ok, I think I've changed that to how you mean. Do feel free to change it further if needed.

Thanks.

josdejong commented:
Thanks for the updates, I like this new feature :)

I've made some changes in the example to make it work on IE11 too.

josdejong commented:
`onCreateMenu` is now available in `v5.30.0` :tada:

RobAley commented:
Thats great, many thanks.
